### PR TITLE
Ocean: Api PoolPair

### PR DESCRIPTION
**Pull request by canonbrother**
## Summary

- [x] list
- [x] get
- [x] token typings
- [x] listpoolswaps (verbose)
- poolpair graph #2855 
  - [x] list paths
  - [x] get best path
  - [x] list swappable tokens

#### TODOs

- ~get token name (from cache)~
- listpoolpaircache (fixed but its disabled, inspect data retrieval on multithread)
- ~getTotalLiquidityUsd (require getBestPath)~
- getAPR (require PoolSwapAggregatedIndexer)
- getUSDVolume (require PoolSwapAggregatedIndexer)
- ~typing: "Bignumber" `String`, eg: i64.to_string() / decimals~
- PoolSwapVerbose.type

#### 🚫 Blocker  

~https://github.com/Jouzo/rust-defichain-rpc/pull/5~

### CODE REVIEW

**Comment by canonbrother** [comment]
checkpoint

File: lib/ain-ocean/src/api/loan.rs [hunk 1]
```diff
@@ -122,7 +122,7 @@ async fn list_collateral_token(
         .into_iter()
         .fake_paginate(&query, skip_while)
         .map(|v| async {
-            let (id, info) = get_token_cached(&ctx, &v.token_id).await?;
+            let (id, info) = get_token_cached(&ctx, &v.token_id).await?.unwrap();
```
Comment (Jouzo) [inline lib/ain-ocean/src/api/loan.rs:125]: Let's `ok_or` instead

File: lib/ain-ocean/src/api/loan.rs [hunk 2]
```diff
@@ -140,7 +140,9 @@ async fn get_collateral_token(
     Extension(ctx): Extension<Arc<AppContext>>,
 ) -> Result<Response<CollateralToken>> {
     let collateral_token = ctx.client.get_collateral_token(token_id).await?;
-    let (id, info) = get_token_cached(&ctx, &collateral_token.token_id).await?;
+    let (id, info) = get_token_cached(&ctx, &collateral_token.token_id)
+        .await?
+        .unwrap();
```
Comment (Jouzo) [inline lib/ain-ocean/src/api/loan.rs:145]: Same as above

File: lib/ain-ocean/src/api/pool_pair/mod.rs [hunk 3]
```diff
@@ -0,0 +1,608 @@
...
+        .0
+        .into_iter()
+        .filter(|(_, p)| !p.symbol.starts_with("BURN-"))
+        .map(|(id, p)| async {
+            let (
+                _,
+                TokenInfo {
+                    name: a_token_name, ..
+                },
+            ) = get_token_cached(&ctx, &p.id_token_a).await?.unwrap();
```
Comment (Jouzo) [inline lib/ain-ocean/src/api/pool_pair/mod.rs:314]: Same as above

File: lib/ain-ocean/src/api/pool_pair/mod.rs [hunk 4]
```diff
@@ -0,0 +1,608 @@
...
+            let (
+                _,
+                TokenInfo {
+                    name: b_token_name, ..
+                },
+            ) = get_token_cached(&ctx, &p.id_token_b).await?.unwrap();
```
Comment (Jouzo) [inline lib/ain-ocean/src/api/pool_pair/mod.rs:320]: Same as above

File: lib/ain-ocean/src/api/pool_pair/mod.rs [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+) -> Result<Response<Option<PoolPairResponse>>> {
+    if let Some((id, pool)) = get_pool_pair_cached(&ctx, id).await? {
+        let total_liquidity_usd = get_total_liquidity_usd(&ctx, &pool).await?;
+        let _apr = get_apr(&ctx, &id, &pool).await?;
+        let (
+            _,
+            TokenInfo {
+                name: a_token_name, ..
+            },
+        ) = get_token_cached(&ctx, &pool.id_token_a).await?.unwrap();
```
Comment (Jouzo) [inline lib/ain-ocean/src/api/pool_pair/mod.rs:355]: Same as above

File: lib/ain-ocean/src/api/pool_pair/mod.rs [hunk 6]
```diff
@@ -0,0 +1,608 @@
...
+        let (
+            _,
+            TokenInfo {
+                name: b_token_name, ..
+            },
+        ) = get_token_cached(&ctx, &pool.id_token_b).await?.unwrap();
```
Comment (Jouzo) [inline lib/ain-ocean/src/api/pool_pair/mod.rs:361]: Same as above

File: lib/ain-ocean/src/api/pool_pair/service.rs [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+
+    let dfi_price_usdt = get_usd_per_dfi(ctx).await?;
+    if dfi_price_usdt.is_zero() {
+        return Ok(dfi_price_usdt);
+    };
+
+    let rewards = if let Some(custom_rewards) = &p.custom_rewards {
+        let total = custom_rewards
+            .iter()
+            .fold(dec!(0), |accumulate: Decimal, account: &String| {
```
Comment (Jouzo) [inline lib/ain-ocean/src/api/pool_pair/service.rs:163]: Should use `try_fold` and return the Result instead of unwrapping here

**Review by Jouzo — APPROVED** [review]

File: lib/ain-ocean/src/api/pool_pair/service.rs (hunk 7 above)
Comment (canonbrother) [reply lib/ain-ocean/src/api/pool_pair/service.rs:163]: haven't got idea to `impl try_fold for Decimal` yet 😓 
replace `unwrap` by `expect` for better error msg temporary... 
https://github.com/DeFiCh/ain/pull/2838/commits/e53119bf45064ce3e4dff0c4e0e4ae4ee27d13cf

**Review by Jouzo — APPROVED** [review]

**Pull request merged**